driver-core/src/main/com/mongodb/operation/OperationHelper.java

```diff
@@ -188,26 +221,54 @@ public void call(final AsyncConnection connection, final Throwable t) {
-                                        final List<? extends WriteRequest> requests, final WriteConcern writeConcern) {
+                                      final List<? extends WriteRequest> requests, final WriteConcern writeConcern,
+                                      final Boolean retryWrites) {
         checkBypassDocumentValidationIsSupported(connection, bypassDocumentValidation, writeConcern);
         validateWriteRequestCollations(connection, requests, writeConcern);
+        validateRetryWrites(connection, retryWrites);
     }
 
     static void validateWriteRequests(final AsyncConnection connection, final Boolean bypassDocumentValidation,
-                                        final List<? extends WriteRequest> requests, final WriteConcern writeConcern,
-                                        final AsyncCallableWithConnection callable) {
+                                      final List<? extends WriteRequest> requests, final WriteConcern writeConcern,
+                                        final Boolean retryWrites, final AsyncCallableWithConnection callable) {
         checkBypassDocumentValidationIsSupported(connection, bypassDocumentValidation, writeConcern, new AsyncCallableWithConnection() {
             @Override
             public void call(final AsyncConnection connection, final Throwable t) {
                 if (t != null) {
                     callable.call(connection, t);
                 } else {
-                    validateWriteRequestCollations(connection, requests, writeConcern, callable);
+                    validateWriteRequestCollations(connection, requests, writeConcern, new AsyncCallableWithConnection() {
+                        @Override
+                        public void call(final AsyncConnection connection, final Throwable t) {
+                            if (t != null) {
+                                callable.call(connection, t);
+                            } else {
+                                validateRetryWrites(connection, retryWrites, callable);
+                            }
+                        }
+                    });
                 }
             }
         });
     }
 
+    static void validateRetryWrites(final Connection connection, final boolean retryWrites) {
+        if (retryWrites && !serverIsAtLeastVersionThreeDotSix(connection.getDescription())) {
+            throw new IllegalArgumentException(format("Retryable writes are not supported by server version: %s",
+                    connection.getDescription().getServerVersion()));
+        }
+    }
+
+    static void validateRetryWrites(final AsyncConnection connection, final Boolean retryWrites,
+                                    final AsyncCallableWithConnection callable) {
+        Throwable throwable = null;
+        if (retryWrites && !serverIsAtLeastVersionThreeDotSix(connection.getDescription())) {
```

Comment:
Just realized something about this check.  There's a corner case where the driver is connected to a 3.6 server but there is no session.  It's rare but could happen in a couple of scenarios:

* A mixed version cluster
* A client configured with more than one MongoCredential (and therefore no implicit session created)

So a better check would be a call to SessionContext#hasSession()

The problem is that at this point in the code we don't know the underlying reason, so the message in the IllegalArgumentException will have to be sufficiently general to point the user in the right direction.

Reply:
A bit more ruminating, and now I think both checks are necessary.  Consider the following timeline:

# topology contains to secondary members of a replica set that are all 3.6 servers
# implicit session is created since all known servers are 3.6
# primary is discovered, and it's a 3.4 server
# The 3.4 server is selected for this write

In that case the driver should also throw an exception

Reply:
When implementing this for libmongoc, I only consider making the write command retryable (i.e. injecting a `txnNumber`) if we have a session to work with (implicit or explicit). See: https://github.com/jmikola/mongo-c-driver/commit/9d92e8e7132eac0fd3e5b5ac69bd9a804c572486#diff-7c720dafff6530406f8a6dd0e76864adR460

In the case of a replica set with mixed 3.6 and 3.4 servers, the cluster would not be capable of running with `{setFeatureCompatibilityVersion: 3.6}`, so write commands would never support a `txnNumber`. The spec already discusses that drivers cannot anticipate if a 3.6+ server has that mode enabled, and must rely on the server to raise an error. Of course, the 3.4 server would be caught preemptively with the wire version check.

Reply:
Also, note that without the hasSession check, the driver will just throw an UnsupportedOperationException a bit later on when it tries to advance the transaction number on a NoOpSessionContext.  But it's better to throw here where there's a bit more context.